File: commands.py
"""
Core module containing all commands which can be invoked by user input. Pass
these classes as parameters to Interpreter class constructors.
"""


class ExplorationCommands(object):
    """
    All commands used in exploration mode.
    """
    def __init__(self, game):
        self.game = game
        # Note that the first matching pattern will be used
        self.patterns = (
            (r"(quit|exit)$",       self.exit),
            (r"l(ook)?$",           self.look),
            (r"i(nv(entory)?)?$",   self.inventory),
            # yes, the long names are actually aliases for the short ones
            (r"e(ast)?$",           self.move_command("e")),
            (r"w(est)?$",           self.move_command("w")),
            (r"n(orth)?$",          self.move_command("n")),
            (r"s(outh)?$",          self.move_command("s")),
            (r"up?$",               self.move_command("u")),
            (r"d(own)?$",           self.move_command("d")),
            (r"(z|wait)$",          self.wait),
        )

    # TODO: better implementation of everything?
    def look(self):
        self.game.writer.write_room(self.game.area.location, "verbose")

    def inventory(self):
        self.game.writer.writecr("Inventory is not yet implemented.")

    def wait(self):
        self.game.writer.writecr("Time passes.")

    def move_command(self, direction):
        def command():
            if self.game.area.move(direction) is not None:
                self.game.writer.write_room(self.game.area.location)
            else:
                self.game.writer.writecr("There is no exit in this direction.")

        return command

    def exit(self):
        self.game.exit_requested = True

File: commands.rb
# Core module containing all commands which can be invoked by user input. Pass
# these classes as parameters to Interpreter class constructors.


# All commands used in exploration mode.
class ExplorationCommands
  def initialize(game)
    @game = game
  end

  def method_missing(symbol, *args)
    case symbol
      when /(quit|exit)$/ then       quit
      when /l(ook)?$/ then           look
      when /i(nv(entory)?)?$/ then   inventory
      when /e(ast)?$/ then           move "e"
      when /w(est)?$/ then           move "w"
      when /n(orth)?$/ then          move "n"
      when /s(outh)?$/ then          move "s"
      when /up?$/ then               move "u"
      when /d(own)?$/ then           move "d"
      when /(z|wait)$/ then          wait
    end
  end

  # TODO: I imagine there is still a better way; I'd like to keep the routing
  # concept, but do something more Rubyish
  private
  def look
    @game.writer.write_room(@game.area.location, "verbose")
  end

  def inventory
    @game.writer.writecr("Inventory is not yet implemented.")
  end

  def wait
    @game.writer.writecr("Time passes.")
  end

  def move(direction)
    if not @game.area.move(direction).nil?
      @game.writer.write_room(@game.area.location)
    else
      @game.writer.writecr("There is no exit in this direction.")
    end
  end

  def quit
    @game.exit_requested = true
  end
end
